File: src/Models/TaskModels.swift
import Foundation

public enum TaskStatus: String, Codable, CaseIterable {
    case needsAction = "needsAction"
    case completed = "completed"
    
    public var isCompleted: Bool {
        return self == .completed
    }
}

public struct TaskList: Identifiable, Hashable, Codable {
    public var id: String
    public var title: String
    public var updated: Date?
    
    public init(id: String = UUID().uuidString, title: String, updated: Date? = Date()) {
        self.id = id
        self.title = title
        self.updated = updated
    }
}

public struct TaskItem: Identifiable, Hashable, Codable {
    public var id: String
    public var listId: String
    public var title: String
    public var notes: String?
    public var due: Date?
    public var status: TaskStatus
    public var completedDate: Date?
    public var parent: String?
    public var position: String?
    public var subtasks: [TaskItem]
    public var updated: Date?
    public var deletedAt: Date?
    
    public init(
        id: String = UUID().uuidString,
        listId: String,
        title: String,
        notes: String? = nil,
        due: Date? = nil,
        status: TaskStatus = .needsAction,
        completedDate: Date? = nil,
        parent: String? = nil,
        position: String? = nil,
        subtasks: [TaskItem] = [],
        updated: Date? = Date(),
        deletedAt: Date? = nil
    ) {
        self.id = id
        self.listId = listId
        self.title = title
        self.notes = notes
        self.due = due
        self.status = status
        self.completedDate = completedDate
        self.parent = parent
        self.position = position
        self.subtasks = subtasks
        self.updated = updated
        self.deletedAt = deletedAt
    }
    
    public var isCompleted: Bool {
        return status == .completed
    }
    
    public var isDeleted: Bool {
        return deletedAt != nil
    }
    
    public var isOverdue: Bool {
        guard let due = due, !isCompleted else { return false }
        return Calendar.current.startOfDay(for: due) < Calendar.current.startOfDay(for: Date())
    }
    
    public var isToday: Bool {
        guard let due = due else { return false }
        return Calendar.current.isDateInToday(due)
    }
    
    public var isUpcoming: Bool {
        guard let due = due, !isCompleted else { return false }
        let now = Date()
        let cal = Calendar.current
        guard let nextWeek = cal.date(byAdding: .day, value: 7, to: now) else { return false }
        let startToday = cal.startOfDay(for: now)
        let dueStart = cal.startOfDay(for: due)
        return dueStart >= startToday && dueStart <= nextWeek
    }
}

public enum SmartFilterType: Hashable, Identifiable, Codable {
    case pending
    case today
    case upcoming
    case calendar
    case all
    case completed
    case trash
    case list(id: String)
    
    public var id: String {
        switch self {
        case .pending: return "smart_pending"
        case .today: return "smart_today"
        case .upcoming: return "smart_upcoming"
        case .calendar: return "smart_calendar"
        case .all: return "smart_all"
        case .completed: return "smart_completed"
        case .trash: return "smart_trash"
        case .list(let id): return "list_\(id)"
        }
    }
    
    public var displayName: String {
        switch self {
        case .pending: return "待辦中"
        case .today: return "今天到期"
        case .upcoming: return "即將到來"
        case .calendar: return "月份日曆"
        case .all: return "全部待辦"
        case .completed: return "已完成事項"
        case .trash: return "垃圾桶"
        case .list: return "清單"
        }
    }
    
    public var iconName: String {
        switch self {
        case .pending: return "circle.dashed"
        case .today: return "star.fill"
        case .upcoming: return "calendar"
        case .calendar: return "calendar.badge.clock"
        case .all: return "tray.full.fill"
        case .completed: return "checkmark.circle.fill"
        case .trash: return "trash"
        case .list: return "list.bullet"
        }
    }
    
    public static func from(id: String) -> SmartFilterType? {
        switch id {
        case "smart_pending": return .pending
        case "smart_today": return .today
        case "smart_upcoming": return .upcoming
        case "smart_calendar": return .calendar
        case "smart_all": return .all
        case "smart_completed": return .completed
        case "smart_trash": return .trash
        default:
            if id.hasPrefix("list_") {
                let lid = String(id.dropFirst(5))
                return .list(id: lid)
            }
            return nil
        }
    }
}

public enum TaskSortOrder: String, CaseIterable, Identifiable {
    case manual = "自訂拖曳順序"
    case dueDate = "到期日優先"
    case creation = "建立順序"
    case title = "標題名稱"
    
    public var id: String { rawValue }
}

File: android/app/src/main/java/com/antigravity/taskcolumn/data/model/TaskModels.kt
package com.antigravity.taskcolumn.data.model

import kotlinx.serialization.Serializable
import java.util.Calendar

@Serializable
enum class TaskStatus(val value: String) {
    NEEDS_ACTION("needsAction"),
    COMPLETED("completed")
}

@Serializable
data class TaskItem(
    var id: String,
    var listId: String,
    var title: String,
    var notes: String? = null,
    var due: Long? = null,
    var status: TaskStatus = TaskStatus.NEEDS_ACTION,
    var parent: String? = null,
    var subtasks: MutableList<TaskItem> = mutableListOf(),
    var updated: Long = System.currentTimeMillis(),
    var deletedAt: Long? = null
) {
    val isCompleted: Boolean
        get() = status == TaskStatus.COMPLETED

    val isTopLevel: Boolean
        get() = parent == null || parent == "null" || parent?.isBlank() == true

    val isOverdue: Boolean
        get() {
            if (isCompleted) return false
            val dueTime = due ?: return false
            val calToday = Calendar.getInstance().apply {
                set(Calendar.HOUR_OF_DAY, 0)
                set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }
            return dueTime < calToday.timeInMillis
        }

    val isToday: Boolean
        get() {
            val dueTime = due ?: return false
            val calDue = Calendar.getInstance().apply { timeInMillis = dueTime }
            val calNow = Calendar.getInstance()
            return calDue.get(Calendar.YEAR) == calNow.get(Calendar.YEAR) &&
                    calDue.get(Calendar.DAY_OF_YEAR) == calNow.get(Calendar.DAY_OF_YEAR)
        }

    val isUpcoming: Boolean
        get() {
            if (isCompleted) return false
            val dueTime = due ?: return false
            val calToday = Calendar.getInstance().apply {
                set(Calendar.HOUR_OF_DAY, 23)
                set(Calendar.MINUTE, 59)
                set(Calendar.SECOND, 59)
                set(Calendar.MILLISECOND, 999)
            }
            return dueTime > calToday.timeInMillis
        }
}

@Serializable
data class TaskList(
    var id: String,
    var title: String,
    var updated: Long = System.currentTimeMillis()
)

sealed class SmartFilterType(val id: String, val title: String) {
    object Pending : SmartFilterType("pending", "待辦中")
    object Today : SmartFilterType("today", "今天到期")
    object Upcoming : SmartFilterType("upcoming", "即將到來")
    object All : SmartFilterType("all", "全部待辦")
    object Completed : SmartFilterType("completed", "已完成事項")
    object Trash : SmartFilterType("trash", "垃圾桶")
    data class CustomList(val listId: String, val listTitle: String) :
        SmartFilterType("list_$listId", listTitle)
}

enum class TaskSortOrder(val displayName: String) {
    MANUAL("自訂排序"),
    DUE_DATE("依到期日"),
    CREATION("依更新時間"),
    TITLE("依標題字母")
}
